Fix import shenanigans 2/2

k4lizen opened the pull request:
based on: https://github.com/pwndbg/pwndbg/pull/3492

+ Renames `aglib/arch.py` into `aglib/arch_mod.py`
+ Renames `aglib/regs.py` into `aglib/regs_mod.py`, now `regs` is a normal class
+ Takes the `class module` out of `aglib/proc.py`, now all functions are global in that file.
+ Refactors `lib/` files to take out any `aglib` mentions (there were 38). (this was a bit involved)
+ Adds documentation that writes down these design decisions.
+ Takes out unnecessary imports from files.
+ Deletes test_find_fastbin_size and the lib/heap/helpers thing.
+ Optimizes lib.cache to use integer flags instead of a dict

k4lizen commented:
~~hmm this is not throwing the "unhashable" error for some stuff where i would expect it to~~

is fine

k4lizen committed:
refactor regs module

k4lizen committed:
add docs about common import pitfalls

k4lizen committed:
make pwndbg.aglib.proc a normal submodule

also make pwndbg.lib.memory not depend on
aglib

k4lizen committed:
prevent import cycle

k4lizen committed:
add comment about importing commands

k4lizen committed:
finish pwndbg.lib.memory fixup

k4lizen committed:
dont use aglib in lib

k4lizen committed:
clearly i need to clarify

k4lizen committed:
fixups all around

k4lizen committed:
a good life

k4lizen committed:
cleanup docs

k4lizen committed:
remove unused imports

k4lizen commented:
<img width="1149" height="632" alt="image" src="https://github.com/user-attachments/assets/d4148416-f934-40ea-bd8a-d1b712277883" />

this is kinda scary hmmmmm

k4lizen committed:
fix nextproginstr

disconnect3d commented on docs/contributing/common-pitfalls.md:
```diff
@@ -0,0 +1,57 @@
+
+#### `from pwndbg.aglib import regs` doesn't work
+
+If you look at `pwndbg/aglib/__init__.py` you will see that `regs` is a None-initialized object and gets swapped out during initialization. To use it properly you must do `aglib.regs.whatever()`.
```
```suggestion
If you look at `pwndbg/aglib/__init__.py` you will see that `regs` is a None-initialized object and gets swapped out during initialization. To use it properly you must not make a copy of it via `from pwndbg.aglib import regs` and instead always access it via `aglib.regs` like: `aglib.regs.whatever()`.
```

disconnect3d committed:
Apply suggestion from @disconnect3d

disconnect3d commented on pwndbg/aglib/__init__.py:
```diff
@@ -11,17 +11,36 @@
+    regs: RegisterManager = cast(RegisterManager, None)
+else:
+    arch = None
+    regs = None
```
I don't like this at all, but I understand we may just need it for now?

Can we fix this somehow in the future?

disconnect3d replied:
Also: wouldn't it be better to default to the default debugger architecture when no file is loaded?

Like in GDB:
```
$ gdb --nx --ex 'show architecture'
The target architecture is set to "auto" (currently "i386").
```

k4lizen replied:
> Also: wouldn't it be better to default to the default debugger architecture when no file is loaded?

the reason i didn't, is because i'm afraid it will cause subtle bugs when
```
from pwndbg.aglib import arch
```
whereas this way, it will just be None and throw an exception.

but I see what you're saying and I will rethink it.

k4lizen replied:
okay so I'm not sure what architecture-specific stuff we can even do before an inferior is running, so I'm not sure if we have a use-case for a default architecture (and as I said, having it as None is more likely to catch bugs)

I agree that if we decide that we do want to set some default architecture here, that we likely want either:

1. The default architecture that the underlying debugger selects, or
2. The architecture the host is on

But in either of those cases, I feel this is a bit out of scope for this PR. So I would prefer to leave it as `= None`, but if you want I can put it back to `= get_pwndbg_architecture("i386")`.

disconnect3d commented on pwndbg/aglib/arch_mod.py:
```diff
@@ -418,7 +418,8 @@ def update() -> None:
 
-    if a.name != pwndbg.aglib.arch.name:
+    # Will be None the first time around.
+    if pwndbg.aglib.arch is None or a.name != pwndbg.aglib.arch.name:
```
Please see https://github.com/pwndbg/pwndbg/pull/3497/changes#r2634812380

disconnect3d commented on pwndbg/aglib/regs_mod.py:
```diff
@@ -117,16 +91,18 @@ def read_reg(self, reg: str, frame: pwndbg.dbg_mod.Frame | None = None) -> int |
+                # FIXME: What is going on here?
+                # if self.cs is None:
+                #     return None
+                # value += self.cs * 16
```
Why do we change the logic here? I believe this was needed for something :)

disconnect3d commented on pwndbg/lib/memory.py:
```diff
@@ -30,11 +30,11 @@ def round_up(address: int, align: int) -> int:
     "Format the given address as a string."
 
-    width = 2 + 2 * pwndbg.aglib.arch.ptrsize
+    width = 2 + 2 * ptrsize
```
Can we really have two pages at the same time with different ptrsize?

Note that we add ptrsize to Page meaning we pass it around lots of times and e.g. if we have 100 pages we now have +100 python integers around

k4lizen replied:
i agree, but how do you propose to resolve the underlying issue of this file calling into aglib?

disconnect3d reviewed (changes requested):
1. Not sure we want to add `ptrsize` to `Page` object, especially in this PR
2. This change also seems to not belog to this PR - it also most likely breaks something:
```py
            if reg == "pc" and pwndbg.aglib.arch.name == "i8086":
                # FIXME: What is going on here?
                # if self.cs is None:
                #     return None
                # value += self.cs * 16
                # pass
```

k4lizen commented:
> This change also seems to not belog to this PR - it also most likely breaks something:

the "what is going on here" was referring to the fact that `regs` doesn't have a `.cs` property, so it would throw an exception at runtime and so i'm not breaking anything by commenting it out

<img width="842" height="194" alt="image" src="https://github.com/user-attachments/assets/eeeb8c1d-9d23-469d-a258-e291f9bf388a" />

i'll change it to this

k4lizen committed:
upd with more pitfalls

k4lizen committed:
fixup from pwndbg.aglib import regs doc

its not really a copy of the object, in this case `regs`
would be bound to the same object as `aglib.regs`
(None) and changes in one would reflect the other
if it wasn't None.

It is a kind of "copy of the name", which is what
you were going for, but maybe this edit makes it
a bit less ambiguous.

k4lizen committed:
fix read_reg(cs)

OBarronCS commented on pwndbg/aglib/regs_mod.py:
```diff
@@ -31,79 +27,58 @@
             if value is None:
                 return None
             value = int(value)
             if reg == "pc" and pwndbg.aglib.arch.name == "i8086":
```
This "pc" should also be changed to "eip", as that's the register GDB uses in this case.

k4lizen replied:
done

k4lizen committed:
pc -> eip. use the pc property if you want the pc!

k4lizen committed:
pitfalls doc imports

k4lizen committed:
fix tests and fix tests

k4lizen commented:
@disconnect3d 
Regarding `ptrsize` to `Page` I don't think the performance is really relevant here but I agree that it is ugly that we are passing a seemingly unrelated field to the constructor of that class (although in practice, everything that uses a Page is in fact allowed to access aglib.arch.ptrsize). There are a few options:

1. Move the whole file to aglib - this is pretty bad, lets not do this
2. Move just the page class to aglib - maybe viable, still feels bad just for one field.
3. Just use 64 bits - I don't think it would be erroneous, but I think it would be noticable for users, and I would rather not make a user-facing sacrifice for a not-that-important implementation detail. 
4. Pass the ptrsize just to the formatting function. I think this is the most reasonable alternative to the current solution, but it is some effort to do. We would have to take `__str__` out of the class and replace it with a custom `string()` function or whatever and change every call-site. It's particularly bad to work on because my editor does not show me `str(pageobject)` references when I ask for references on `__str__`, so I would rather not do it for this PR.

I will note that the problem of a lib file accessing aglib (which I solve here by passing ptrsize) is not a theoretical one, I have gotten import cycles due to this during this refactor.

k4lizen committed:
flew too close to the sun

k4lizen committed:
okay I see the issue

k4lizen committed:
remove duplicated gdb test

k4lizen committed:
fix lldb by setting proper event priorities

so satisfying

k4lizen committed:
remove unneeded gdb test

again

k4lizen committed:
move this troublesome line to PR 3/2
